### PR TITLE
8264400: (fs) WindowsFileStore equality depends on how the FileStore was constructed

### PR DESCRIPTION
**Pull request by bplb**
Please consider this change to make `sun.nio.fs.WindowsFileStore.equals()` return `true` if the root strings of the two objects are equal under case insensitive comparison, and the two `WindowsFileStore`s have the same volume information.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264400](https://bugs.openjdk.java.net/browse/JDK-8264400): (fs) WindowsFileStore equality depends on how the FileStore was constructed

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to e1d7b21a15b08534ad4b0ed0453b08f750f3babb

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3279/head:pull/3279` \
`$ git checkout pull/3279`

Update a local copy of the PR: \
`$ git checkout pull/3279` \
`$ git pull https://git.openjdk.java.net/jdk pull/3279/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3279`

View PR using the GUI difftool: \
`$ git pr show -t 3279`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3279.diff">https://git.openjdk.java.net/jdk/pull/3279.diff</a>

</details>

### CODE REVIEW

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 1]
```diff
@@ -232,7 +232,15 @@ public boolean equals(Object ob) {
...
-        return root.equals(other.root);
+        if (root.equals(other.root))
+            return true;
+        if (root.equalsIgnoreCase(other.root)) {
+             VolumeInformation otherInfo = other.volumeInformation();
+            return volInfo.volumeName().equals(otherInfo.volumeName()) &&
+                volInfo.fileSystemName().equals(otherInfo.fileSystemName()) &&
+                volInfo.volumeSerialNumber() == otherInfo.volumeSerialNumber();
+        }
+        return false;
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: The volType == DRIVE_REMOTE case is the reason why we didn't fix this one in the past. Can you summarise that the VolumeInfomation is for this case?

Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: I was not able to test that case yet. Using the VolumeInformation was intended to limit spurious equality.

Can you explain abut DRIVE_REMOTE?

Comment (AlanBateman) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: > I was not able to test that case yet. Using the VolumeInformation was intended to limit spurious equality.
> 
> Can you explain abut DRIVE_REMOTE?

I think the volume type will be DRIVE_REMOTE when the root is a UNC (the constants for the volume types are defined in WIndowsConstants). I'm not 100% sure if the server and share names should be compared with or without regard to case so I think we have a bit more exploration to do there. The DRIVE_FIXED case is clear, and it would be good to at least get that fixed as there has been C:\ vs. c:\ confusion in a number of bug reports.

Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: Based on testing in a Win 10 VM, the volume type is in fact DRIVE_REMOTE. The volume serial number however is zero. From what I can test via testing and reading the server and share names don't care about case.

Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: If we are unsure, for UNC, perhaps the server and share name comparisons can remain case sensitive but the file path component comparison be case insensitive?

Comment (AlanBateman) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: Maybe this is something that the Microsoft folks can help us with, it might be that it depends on the transport.

Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:243]: Good idea. I will ask them.

**Comment by bplb** [comment]
I already have this information.

1. `C:\` - Win10 system drive

`Root: C:\
volName: System, fsName: NTFS, volSN: 1620920694, volType: 3`

2. `E:\` - CD drive (VirtualBox Guest Additions)

`Root: E:\
volName: VBox_GAs_6.1.18, fsName: CDFS, volSN: -1657170319, volType: 5`

3. `Z:\` - VirtualBox shared folder

`Root: Z:\
volName: VBOX_VirtualBox, fsName: VBoxSharedFolderFS, volSN: 16777221, volType:
4`

4. `\\BPB\Public` - Local share created for `C:\Users\Public`

`Root: \\BPB\pUBLIC\
volName: System, fsName: NTFS, volSN: 1620920694, volType: 4`

5. `\\192.168.0.111\Shared\` - "Remote" share of macOS host `/Users/Shared`

`Root: \\192.168.0.111\Shared\
volName: Macintosh HD - Data, fsName: NTFS, volSN: 0, volType: 4`

**Comment by bplb** [comment]
I already tested with other cases but didn't include it:

```
Root: \\bpb\Public\
volName: System, fsName: NTFS, volSN: 1620920694, volType: 4

Root: \\BPB\pUBLIC\
volName: System, fsName: NTFS, volSN: 1620920694, volType: 4

Root: \\192.168.0.111\Shared\
volName: Macintosh HD - Data, fsName: NTFS, volSN: 0, volType: 4

Root: \\192.168.0.111\shared\
volName: Macintosh HD - Data, fsName: NTFS, volSN: 0, volType: 4
```

Good point about `hashCode()`.

**Comment by bplb** [comment]
If we want to constrain the case insensitive comparison to fixed drives, then this might work:
```
    @Override
    public boolean equals(Object ob) {
        if (ob == this)
            return true;
        if (!(ob instanceof WindowsFileStore))
            return false;
        WindowsFileStore other = (WindowsFileStore)ob;
        if (root.equals(other.root))
            return true;
        if (volType == DRIVE_FIXED && other.volumeType() == DRIVE_FIXED)
            return root.equalsIgnoreCase(other.root);
        return false;
    }

    @Override
    public int hashCode() {
        return volType == DRIVE_FIXED ?
            root.toLowerCase().hashCode() : root.hashCode();
    }
```

**Comment by AlanBateman** [comment]
> If we want to constrain the case insensitive comparison to fixed drives, then this might work:

Yes, although it would be good to extend this to the DRIVE_REMOTE and other cases too, but that requires being confident that it is correct (and I'm not sure we have enough information to say this yet).

An initial change for the DRIVE_FIXED case is okay. It will mean using Locale.ROOT or Locale.US. Also the hash code will need to be cached.

**Comment by bplb** [comment]
What do you intend by the `Locale.X` comment?

**Comment by bplb** [comment]
(Commit message for 9260dfd intended `DRIVE_FIXED`, not `DRIVE_LOCAL`: inadvertently conflated with `MNT_LOCAL` on `macOS`.)

**Comment by bplb** [comment]
(Commit message for 9260dfd intended `DRIVE_FIXED`, not `DRIVE_LOCAL`: inadvertently conflated with `MNT_LOCAL` on `macOS`.)

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 2]
```diff
@@ -44,6 +52,9 @@
     private final int volType;
     private final String displayName;   // returned by toString
 
+    private boolean hasHashCode = false; // as hashCode can be any int
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:55]: This is buggy. It would be simpler to drop hasHashCode and just race when hash is 0.

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 3]
```diff
@@ -221,8 +231,7 @@ public boolean supportsFileAttributeView(String name) {
         if (name.equals("owner"))
             return supportsFileAttributeView(FileOwnerAttributeView.class);
         if (name.equals("user"))
-            return supportsFileAttributeView(UserDefinedFileAttributeView.class);
-        return false;
+            return supportsFileAttributeView(UserDefinedFileAttributeView.class); return false;
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:234]: Can you revert this change, it's not part of the fix and it makes this method harder to maintain.

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 4]
```diff
@@ -232,12 +241,20 @@ public boolean equals(Object ob) {
         if (!(ob instanceof WindowsFileStore))
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:241]: We can change this to
if (ob instanced WindowsFileStore other) { ... } and it will avoid the cast.

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 5]
```diff
@@ -232,12 +241,20 @@ public boolean equals(Object ob) {
...
+        if (volType == DRIVE_FIXED && other.volumeType() == DRIVE_FIXED)
+            return root.equalsIgnoreCase(other.root);
+        return false;
     }
 
     @Override
     public int hashCode() {
-        return root.hashCode();
+        if (hashCode == 0) { // Don't care about race
+            hashCode = volType == DRIVE_FIXED ?
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:254]: minor nit but would be a bit easier to read if changed to 
hashCode = (volType == DRIVE_FIXED) ? ...

File: test/jdk/java/nio/file/FileStore/Basic.java [hunk 6]
```diff
@@ -80,6 +80,16 @@ static void doTests(Path dir) throws IOException {
         assertTrue(store2.equals(store1));
         assertTrue(store1.hashCode() == store2.hashCode());
 
+        if (Platform.isWindows()) {
+            /**
+             * Test: FileStore.equals() should not be case sensitive
+             */
+            FileSystem fs = FileSystems.getDefault();
+            FileStore upper = Files.getFileStore(fs.getPath("C:\\"));
+            FileStore lower = Files.getFileStore(fs.getPath("c:\\"));
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/FileStore/Basic.java:89]: Path.of("c:\\") can be used if you want.

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java (hunk 3 above)
Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:234]: I did revert it before but I guess it crept back in ... sigh.

File: test/jdk/java/nio/file/FileStore/Basic.java (hunk 6 above)
Comment (bplb) [reply test/jdk/java/nio/file/FileStore/Basic.java:89]: Will simplify to that in final commit.

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 7]
```diff
@@ -44,6 +52,8 @@
     private final int volType;
     private final String displayName;   // returned by toString
 
+    private int hashCode = 0;
+
```
Comment (AlanBateman) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:56]: Good, I think this is safe and equals/hashCode are consistent. I suspect we can extend this to other (maybe all) volume type but it needs a bit more exploration. Minor nit is that you can't need to initialise hashCode to 0.

Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:56]: I think this issue was specific to the `DRIVE_FIXED` case so further enhancement can be done under another issue once the information becomes knowd.

Comment (AlanBateman) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:56]: > I think this issue was specific to the `DRIVE_FIXED` case so further enhancement can be done under another issue once the information becomes knowd.

That's okay with me.

File: test/jdk/java/nio/file/FileStore/Basic.java [hunk 8]
```diff
@@ -80,6 +80,16 @@ static void doTests(Path dir) throws IOException {
         assertTrue(store2.equals(store1));
         assertTrue(store1.hashCode() == store2.hashCode());
 
+        if (Platform.isWindows()) {
+            /**
+             * Test: FileStore.equals() should not be case sensitive
+             */
+            FileSystem fs = FileSystems.getDefault();
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/FileStore/Basic.java:87]: I assume this line is no longer needed.

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java [hunk 9]
```diff
@@ -229,15 +239,22 @@ public boolean supportsFileAttributeView(String name) {
...
     }
 
     @Override
     public int hashCode() {
-        return root.hashCode();
+        if (hashCode == 0) { // Don't care about race
+            hashCode = (volType == DRIVE_FIXED) ?
+                root.toLowerCase(Locale.ROOT).hashCode() : root.hashCode();
+        }
+        return hashCode;
```
Comment (turbanoff) [inline src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:257]: As I know. to make data race benign, field should be read only once. But there is second read which can still return `0`

Comment (AlanBateman) [reply src/java.base/windows/classes/sun/nio/fs/WindowsFileStore.java:257]: You can are right, missed in the final review (as this one went through many iterations).
